### PR TITLE
Introduce google_gax package

### PR DESCRIPTION
**Pull request by chingor13**
This common library will contain the shared library code which will allow us to simplify the generator and more easily update libraries without having to rebuild and republish every package individually.

### CODE REVIEW

File: clients/gax/lib/google_api/gax.ex [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule Gax do
```
Comment (dazuma) [inline clients/gax/lib/google_api/gax.ex]: Should this be GoogleApi.Gax?
Do we even need this?

Comment (chingor13) [reply clients/gax/lib/google_api/gax.ex]: We don't need it. I'll remove.

File: clients/gax/lib/google_api/gax.ex [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2018 Google Inc.
```
Comment (dazuma) [inline clients/gax/lib/google_api/gax.ex]: Copyright for new files should be:
```
# Copyright 2018 Google LLC
```

File: clients/gax/lib/google_api/gax/request.ex [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  - definitions (Map) - Map of parameter name to parameter location.
+  - options (KeywordList) - The provided optional parameters
+
+  ## Returns
+
+  Map
+  """
+  @spec add_optional_params(
+          GoogleApi.Gax.Request.t(),
+          %{optional(atom()) => param_location},
```
Comment (dazuma) [inline clients/gax/lib/google_api/gax/request.ex]: I'm noticing you have parens `()` on some typespecs and not others. Most have parens but a few don't, such as `param_location` here and a few lines down.

Comment (chingor13) [reply clients/gax/lib/google_api/gax/request.ex]: Fixed.

File: clients/gax/lib/google_api/gax.ex (hunk 2 above)
Comment (chingor13) [reply clients/gax/lib/google_api/gax.ex]: Fixed

File: clients/gax/lib/google_api/gax/request.ex [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  @type t :: %__MODULE__{
+          method: method(),
+          url: String.t(),
+          body: keyword(),
+          query: keyword(),
+          file: keyword(),
+          header: keyword()
+        }
+
+  @spec new() :: map()
```
Comment (dazuma) [inline clients/gax/lib/google_api/gax/request.ex]: Should this be `new() :: GoogleApi.Gax.Request.t()`?

**Pull request merged**